### ISSUE TITLE
Monitoring tab shows wrong CPU utilization

### ISSUE THREAD

**Issue by simitt** [elastic/beats#9377]
Testing with `metricbeat 6.5.1` I see a very low CPU utilization, although the system is running on ~80% on all cores.

![screen shot 2018-12-04 at 14 15 28](https://user-images.githubusercontent.com/5555349/49444409-659ed880-f7cf-11e8-9200-ff7e78ea83a4.png)

**Comment by ycombinator** [comment]
I tried running metricbeat with monitoring turned on and compared the CPU utilization chart with the output of `top`. The numbers seem to be in the same ballpark for me.

@simitt Could you include the steps you took to cause this on your end? Perhaps that'll help me reproduce it on my end. Thanks!

**Comment by simitt** [comment]
I used 6.5.2 for testing and basically didn't do anything special. For generating load on the beat itself, I set `period: 10ms` in `modules.d/system.yml` and in `modules.d/golang.yml`.

On my local machine with 4 CPUs this results in:
*htop* (values between 95 and 160%)
![screen shot 2018-12-11 at 13 26 06](https://user-images.githubusercontent.com/5555349/49801216-df900e00-fd49-11e8-9cbc-dd7c2a5eed03.png)

*monitoring*:
![screen shot 2018-12-11 at 13 32 56](https://user-images.githubusercontent.com/5555349/49801136-a6f03480-fd49-11e8-8b05-8bc5db65eaeb.png)

I also used it in a cloud gcp setup, with a small machine of *1CPU*.  It doesn't seem to not normalize the amount of used CPU to the available CPU for my box.
*htop* 
![screen shot 2018-12-11 at 13 32 09](https://user-images.githubusercontent.com/5555349/49801204-d606a600-fd49-11e8-8d3c-256fa00c8b2b.png)

*monitoring*:
![screen shot 2018-12-11 at 13 38 05](https://user-images.githubusercontent.com/5555349/49801273-077f7180-fd4a-11e8-8ceb-a0193746e24b.png)

**Comment by ycombinator** [comment]
That's perfect, @simitt. I'm able to reproduce it locally now and can investigate further. Thanks!

**Comment by ycombinator** [comment]
I'm not sure what's going on with the calculation for this metric in the UI code (yet) but I started by looking at the raw documents in `.monitoring-beats-6-*` to try and narrow down where the problem might be.

The metric is being calculated from the `beats_stats.metrics.beat.cpu.total.value` field, which is a counter metric and whose units are `ms`. 

When the time picker in the UI is set to "Last 1 hour", the bucket size of each X-axis bucket gets automatically set to `30s`. So I setup a `date_histogram` aggregation with an `interval` of `30s`. Since we're dealing with a counter metric, I asked for the `max` of the counter metric field within each date histogram bucket. Also, for the same reason, I asked for a derivative (difference) between adjacent date histogram bucket values for the metric, so we get to see the CPU utilization (in ms) just for each 30s time period. Finally, for ease of inspection, I restricted the time range to the last couple of minutes. Here's the final query with its results:

#### Query
```
POST .monitoring-beats-6-*/_search?filter_path=aggregations
{
  "size": 0,
  "query": {
    "range": {
      "timestamp": {
        "gte": "now-2m/m",
        "lte": "now"
      }
    }
  },
  "aggs": {
    "timeseries": {
      "date_histogram": {
        "field": "timestamp",
        "interval": "30s"
      },
      "aggs": {
        "total_cpu_ms": {
          "max": {
            "field": "beats_stats.metrics.beat.cpu.total.value"
          }
        },
        "rate": {
          "derivative": {
            "buckets_path": "total_cpu_ms"
          }
        }
      }
    }
  }
}
```

#### Results
```json
{
  "aggregations" : {
    "timeseries" : {
      "buckets" : [
        {
          "key_as_string" : "2018-12-11T18:36:00.000Z",
          "key" : 1544553360000,
          "doc_count" : 4,
          "total_cpu_ms" : {
            "value" : 6787848.0
          }
        },
        {
          "key_as_string" : "2018-12-11T18:36:30.000Z",
          "key" : 1544553390000,
          "doc_count" : 3,
          "total_cpu_ms" : {
            "value" : 6857067.0
          },
          "rate" : {
            "value" : 69219.0
          }
        },
        {
          "key_as_string" : "2018-12-11T18:37:00.000Z",
          "key" : 1544553420000,
          "doc_count" : 4,
          "total_cpu_ms" : {
            "value" : 6925499.0
          },
          "rate" : {
            "value" : 68432.0
          }
        },
        {
          "key_as_string" : "2018-12-11T18:37:30.000Z",
          "key" : 1544553450000,
          "doc_count" : 3,
          "total_cpu_ms" : {
            "value" : 6994194.0
          },
          "rate" : {
            "value" : 68695.0
          }
        }
      ]
    }
  }
}
```

If you take the `rate.value` from any date histogram bucket and divide it by the bucket size (in the same unit, `ms`), we get the % of CPU utilized during that date histogram bucket. For example, if you take the very last bucket in the above results, you get the % CPU utilization for that bucket = `68695.0 / (30 * 1000) = 2.28 = 228%`. At this point, when I flipped over to my `htop` output, I see something very close to it:

<img width="1601" alt="screen shot 2018-12-11 at 10 41 49 am" src="https://user-images.githubusercontent.com/51061/49822280-874d1200-fd31-11e8-8bb5-ae428d13052a.png">

In conclusion, it appears that the data in the `.monitoring-beats-6-*` docs is good; we are just not performing the right calculation in the UI code.

**Comment by ycombinator** [comment]
FYI @chrisronline, since this appears to be a UI bug. Let me know if you want a separate issue in the Kibana repo but there's information in this one that you might want to look at as well.

**Comment by simitt** [comment]
Thank you for the detailed analyzation! We do have the same problem for the CPU utilization in the APM monitoring. Let me know if I should create a separate issue for that (but I assume it is the same bug).

**Comment by ycombinator** [comment]
It's the same bug, this issue should suffice for both.

**Comment by chrisronline** [comment]
It seems the math is just wrong here: https://github.com/elastic/kibana/blob/master/x-pack/plugins/monitoring/server/lib/metrics/beats/classes.js#L155 and https://github.com/elastic/kibana/blob/master/x-pack/plugins/monitoring/server/lib/metrics/apm/classes.js#L72

**Comment by ycombinator** [comment]
@chrisronline I think this issue can be closed now, as it was resolved via elastic/kibana#27416?

**Comment by chrisronline** [comment]
Yup. Closing

**Comment by ycombinator** [comment]
Hey @chrisronline, in your [comment above](https://github.com/elastic/beats/issues/9377#issuecomment-446705461), you said:

> It seems the math is just wrong here: elastic/kibana:x-pack/plugins/monitoring/server/lib/metrics/beats/classes.js@master#L155 and elastic/kibana:x-pack/plugins/monitoring/server/lib/metrics/apm/classes.js@master#L72

Specifically, you linked to beats and apm code.

However, looking at the [PR that resolved this issue](https://github.com/elastic/kibana/pull/27416), it appears on the beats code was fixed. Is there a corresponding PR for the APM code as well?

**Comment by chrisronline** [comment]
@ycombinator There is not. Good catch. Filed https://github.com/elastic/kibana/issues/33025

**Comment by simitt** [comment]
Excerpt from [my comment](https://github.com/elastic/beats/issues/9377#issuecomment-446189353) above:

> I also used it in a cloud gcp setup, with a small machine of 1CPU. It doesn't seem to not normalize the amount of used CPU to the available CPU for my box.

@ycombinator I assumed the CPU usage is supposed to be shown relative to the CPU ticks assigned to the container and not relative to the system CPU. I am not sure anymore, could you please clarify.

**Comment by blakerouse** [comment]
Closing as the fix landed in elastic/kibana#37230
